File: payment-service/src/main/java/com/study/payment/application/service/PaymentService.java
package com.study.payment.application.service;

import com.study.payment.application.dto.CreatePaymentRequest;
import com.study.payment.application.dto.PaymentResponse;
import com.study.payment.domain.entity.Payment;
import com.study.payment.domain.repository.PaymentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final KafkaTemplate<String, String> kafkaTemplate;

    @Transactional
    public PaymentResponse createPayment(Long userId, CreatePaymentRequest request) {
        Payment payment = Payment.builder()
                .bookingId(request.getBookingId())
                .userId(userId)
                .amount(request.getAmount())
                .build();
        Payment saved = paymentRepository.save(payment);
        kafkaTemplate.send("payment-events", "PAYMENT_CREATED:" + saved.getId());
        return PaymentResponse.from(saved);
    }

    public PaymentResponse getPayment(Long id) {
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Payment not found: " + id));
        return PaymentResponse.from(payment);
    }

    @Transactional
    public PaymentResponse completePayment(Long id) {
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Payment not found: " + id));
        payment.complete();
        kafkaTemplate.send("payment-events", "PAYMENT_COMPLETED:" + id);
        return PaymentResponse.from(payment);
    }

    @Transactional
    public PaymentResponse refundPayment(Long id) {
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Payment not found: " + id));
        payment.refund();
        kafkaTemplate.send("payment-events", "PAYMENT_REFUNDED:" + id);
        return PaymentResponse.from(payment);
    }
}

File: payment-service/src/main/kotlin/com/study/payment/application/service/PaymentService.kt
package com.study.payment.application.service

import com.study.common.event.PaymentCompletedEvent
import com.study.common.event.PaymentFailedEvent
import com.study.common.event.PaymentRefundedEvent
import com.study.common.exception.EntityNotFoundException
import com.study.common.outbox.OutboxEventPublisher
import com.study.payment.application.dto.CreatePaymentRequest
import com.study.payment.application.dto.PaymentResponse
import com.study.payment.domain.entity.Payment
import com.study.payment.domain.repository.PaymentRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class PaymentService(
    private val paymentRepository: PaymentRepository,
    private val outboxEventPublisher: OutboxEventPublisher
) {
    @Transactional
    fun createPayment(userId: Long, request: CreatePaymentRequest): PaymentResponse {
        val bookingId = requireNotNull(request.bookingId) { "Booking ID is required" }
        val amount = requireNotNull(request.amount) { "Amount is required" }
        val payment = Payment(bookingId = bookingId, userId = userId, amount = amount)
        val saved = paymentRepository.save(payment)
        return PaymentResponse.from(saved)
    }

    fun getPayment(id: Long): PaymentResponse =
        PaymentResponse.from(paymentRepository.findById(id).orElseThrow { EntityNotFoundException("Payment", id) })

    fun getPaymentsByUser(userId: Long): List<PaymentResponse> =
        paymentRepository.findByUserId(userId).map { PaymentResponse.from(it) }

    @Transactional
    fun completePayment(id: Long): PaymentResponse {
        val payment = paymentRepository.findById(id).orElseThrow { EntityNotFoundException("Payment", id) }
        payment.complete()
        outboxEventPublisher.publish("payment-events", PaymentCompletedEvent(payment.id, payment.bookingId, payment.userId, payment.amount))
        return PaymentResponse.from(payment)
    }

    @Transactional
    fun refundPayment(id: Long): PaymentResponse {
        val payment = paymentRepository.findById(id).orElseThrow { EntityNotFoundException("Payment", id) }
        payment.refund()
        outboxEventPublisher.publish("payment-events", PaymentRefundedEvent(payment.id, payment.bookingId, payment.userId, payment.amount))
        return PaymentResponse.from(payment)
    }

    @Transactional
    fun failPayment(id: Long, reason: String): PaymentResponse {
        val payment = paymentRepository.findById(id)
            .orElseThrow { EntityNotFoundException("Payment", id) }
        payment.fail()
        outboxEventPublisher.publish("payment-events",
            PaymentFailedEvent(payment.id, payment.bookingId, payment.userId, reason))
        return PaymentResponse.from(payment)
    }
}
